### ISSUE TITLE
wait_for_connection with delegate_to does not use delegated-to host's connection parameters

### ISSUE THREAD

**Issue by nkakouros** [ansible/ansible#36519]
##### ISSUE TYPE

 - Feature Idea

##### COMPONENT NAME
wait_for_connection

##### ANSIBLE VERSION

```
ansible 2.6.0 (devel 8743b20c8e) last updated 2018/02/13 17:38:47 (GMT +200)
  config file = /home/nikos/Projects/ansible-demo/ansible.cfg
  configured module search path = ['/home/nikos/Projects/ansible-demo/ansible/lib/modules']
  ansible python module location = /home/nikos/Projects/ansible/lib/ansible
  executable location = /home/nikos/Projects/ansible/bin/ansible
  python version = 3.6.4 (default, Jan  5 2018, 02:35:40) [GCC 7.2.1 20171224]
```

##### CONFIGURATION
```
ANSIBLE_FORCE_COLOR(/home/nikos/Projects/ansible-demo/ansible.cfg) = True
ANSIBLE_NOCOWS(/home/nikos/Projects/ansible-demo/ansible.cfg) = True
ANSIBLE_PIPELINING(/home/nikos/Projects/ansible-demo/ansible.cfg) = True
ANY_ERRORS_FATAL(/home/nikos/Projects/ansible-demo/ansible.cfg) = True
COLOR_DEBUG(/home/nikos/Projects/ansible-demo/ansible.cfg) = magenta
DEFAULT_CALLBACK_PLUGIN_PATH(/home/nikos/Projects/ansible-demo/ansible.cfg) = ['/home/nikos/Projects/ansible-demo/ansible/lib/plugins/callback']
DEFAULT_CALLBACK_WHITELIST(/home/nikos/Projects/ansible-demo/ansible.cfg) = ['timer', 'log_plays_overwrite']
DEFAULT_EXECUTABLE(/home/nikos/Projects/ansible-demo/ansible.cfg) = /bin/bash
DEFAULT_FILTER_PLUGIN_PATH(/home/nikos/Projects/ansible-demo/ansible.cfg) = ['/home/nikos/Projects/ansible-demo/ansible/lib/plugins/filter']
DEFAULT_GATHERING(/home/nikos/Projects/ansible-demo/ansible.cfg) = explitic
DEFAULT_GATHER_SUBSET(/home/nikos/Projects/ansible-demo/ansible.cfg) = !hardware,!ohai,!facter
DEFAULT_HOST_LIST(/home/nikos/Projects/ansible-demo/ansible.cfg) = ['/home/nikos/Projects/ansible-demo/inventory.ini']
DEFAULT_INVENTORY_PLUGIN_PATH(/home/nikos/Projects/ansible-demo/ansible.cfg) = ['/home/nikos/Projects/ansible-demo/ansible/lib/plugins/inventory']
DEFAULT_LOOKUP_PLUGIN_PATH(/home/nikos/Projects/ansible-demo/ansible.cfg) = ['/home/nikos/Projects/ansible-demo/ansible/lib/plugins/lookup']
DEFAULT_MANAGED_STR(/home/nikos/Projects/ansible-demo/ansible.cfg) = Ansible managed (do not edit, changes may be overwritten)
DEFAULT_MODULE_PATH(/home/nikos/Projects/ansible-demo/ansible.cfg) = ['/home/nikos/Projects/ansible-demo/ansible/lib/modules']
DEFAULT_MODULE_UTILS_PATH(/home/nikos/Projects/ansible-demo/ansible.cfg) = ['/home/nikos/Projects/ansible-demo/ansible/lib/module_utils']
DEFAULT_ROLES_PATH(/home/nikos/Projects/ansible-demo/ansible.cfg) = ['/home/nikos/Projects/ansible-demo/roles']
DEFAULT_STDOUT_CALLBACK(/home/nikos/Projects/ansible-demo/ansible.cfg) = dense_with_book
DEFAULT_UNDEFINED_VAR_BEHAVIOR(/home/nikos/Projects/ansible-demo/ansible.cfg) = True
DEFAULT_VARS_PLUGIN_PATH(/home/nikos/Projects/ansible-demo/ansible.cfg) = ['/home/nikos/Projects/ansible-demo/ansible/lib/plugins/vars']
DEPRECATION_WARNINGS(/home/nikos/Projects/ansible-demo/ansible.cfg) = False
DISPLAY_ARGS_TO_STDOUT(/home/nikos/Projects/ansible-demo/ansible.cfg) = True
DISPLAY_SKIPPED_HOSTS(/home/nikos/Projects/ansible-demo/ansible.cfg) = False
HOST_KEY_CHECKING(/home/nikos/Projects/ansible-demo/ansible.cfg) = False
RETRY_FILES_ENABLED(/home/nikos/Projects/ansible-demo/ansible.cfg) = False
USE_PERSISTENT_CONNECTIONS(/home/nikos/Projects/ansible-demo/ansible.cfg) = True
```

Arch Linux latest

##### SUMMARY
When running a `wait_for_connection` from a linux host task with delegation to a windows host, the connection that is being tried is an ssh one.

##### STEPS TO REPRODUCE

```yaml
- add_host: &winrm
    hostname: win_instance
    ansible_host: "{{ win_instance.instance_data[0].public_ip }}"
    ansible_connection: winrm
    ansible_port: 5986
    ansible_winrm_server_cert_validation: ignore
    ansible_winrm_transport: ntlm
    ansible_user: "{{ win_server.user }}"
    ansible_password: "{{ win_instance_password.stdout_lines[1] |
    ⮱regex_replace('password:[\\s]*([^s]*)', '\\1') }}"
- gce_firewall:
    name: custom-windows-image
    network: "{{ project_facts.network.name }}"
    allowed: udp:3389;tcp:3389;tcp:5986;icmp
    src_ranges: "{{ project_facts.local.address }}/32"
    target_tags: custom-windows-image
- wait_for_connection:
  delegate_to: custom-windows-image
  # vars: *winrm
```

##### EXPECTED RESULTS
WinRM should be used to connect to the windows host using the information from the inventory.

##### ACTUAL RESULTS

```
TASK [wait_for_connection ] *****************************************************************************************************************************************************************
task path: ....gce-setup.yml:68
wait_for_connection: attempting ping module test
Using module file /home/nikos/Projects/ansible/lib/ansible/modules/system/ping.py
<custom-windows-image> ESTABLISH SSH CONNECTION FOR USER: None
<custom-windows-image> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o 'IdentityFile="....../ssh_key"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/nikos/.ansible/cp/14627d6ae4 custom-windows-image '/bin/bash -c '"'"'GCE_PROJECT=en2720-2017 GCE_CREDENTIALS_FILE_PATH='"'"'"'"'"'"'"'"'.....auth.json'"'"'"'"'"'"'"'"' GCE_PEM_FILE_PATH='"'"'"'"'"'"'"'"'.....auth.json'"'"'"'"'"'"'"'"' GCE_EMAIL=e@e.iam.gserviceaccount.com /usr/bin/python && sleep 0'"'"''
wait_for_connection: attempting ping module test
...
```
and so on.

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/modules/utilities/logic/wait_for_connection.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/utilities/logic/wait_for_connection.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @dagwieers
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by dagwieers** [comment]
Well, **wait_for_connection** is an action plugin, so it cannot be delegated. It always runs from the Ansible master node. But if you expected it to test another remote_host, you can influence this using `vars`. Or create a new play targeting your host and run `wait_for_connection` there.

**Comment by nkakouros** [comment]
[**@dagwieers**](https://github.com/dagwieers) commented on [Feb 21, 2018, 4:27 PM GMT+1](https://github.com/ansible/ansible/issues/36519#issuecomment-367362879 "2018-02-21T15:27:02Z - Replied by Github Reply Comments"):
> **wait_for_connection** is an action plugin, so it cannot be delegated

I think this should be mentioned somewhere. I created #36524 to add an example on the module's docs.

BTW, is it out of scope to ask the addition of a parameter in the `wait_for_connection` module that specifies which host to wait_for explicitly? As it is now, it waits for the current host the playbook is run against. Using `vars` to influence this is a bit unintuitive. Perhaps:

```yaml
- wait_for_connection:
    host: inventory_host_2
```

**Comment by nkakouros** [comment]
BTW, there is also no warning that `delegate_to` is not supported by the module.

**Comment by dagwieers** [comment]
It is an action plugin, not an actual module.

That said, we could support `delegate_to` one way or the other, but as I already indicated you don't need to use `delegate_to` in your example. Just target your newly added nodes in a new play and it will actually wait for all your hosts to come online in parallel.

So in my humble opinion `delegate_to` support is not really required.

**Comment by nkakouros** [comment]
OK. 

I don't want to waste your time, but I think there is a UX issue here. How is a user supposed to know that delegate_to and other properties are not used by action plugins? I did not find any relevant information in the documentation (eg [this](http://docs.ansible.com/ansible/devel/plugins/action.html) says nothing about the issue). Neither do plugins' docs say anything about that (eg [here](http://docs.ansible.com/ansible/devel/modules/wait_for_connection_module.html)). They are in fact called _Utility **Modules**_. Nor is Ansible emitting a warning when delegate_to, etc are used with action plugins. 

I do understand the technical reasoning for not supporting "module properties", but shouldn't this be communicated to the users somehow?

**Comment by dagwieers** [comment]
@tterranigma I agree, the module documentation is not very clear about the distinction. I can only assume this is done to not expose this kind of complexity to (new) users. As some modules consist of an action-plugin *and* a remotely-executed module, only a remotely-executed module (what most people conceptually understand), and some are only an action-plugin, and all this affects how tasks work.

Maybe best to open a new issue related to this so others can voice their opinion on this. I personally wouldn't mind making it more clear from the documentation, maybe from the module notes ?

Again, in your case the proper solution is to target your (newly created) system in a subsequent play where you can use **wait_for_connection** without the need for delegation. Just as you would for any subsequent action to target that newly available system.

**Comment by nkakouros** [comment]
@dagwieers Thanks for your suggestions and advice. I will create a new issue for the documentation part.

**Comment by dagwieers** [comment]
I will keep this issue open because we may still implement `delegate_to` support as technically this is possible. I just don't know if this is desirable as it is counter-intuitive IMO.

**Comment by bcoca** [comment]
proposal for adding info to modules, including 'what keywords are not supported' https://github.com/ansible/proposals/issues/68

@dagwieers depending on action plugin, delegate_to is supported i.e copy/assemble/syncronize/etc ... for others it makes no sense i.e debug, add_host, etc

**Comment by dagwieers** [comment]
@bcoca Well, those you list have modules, so there's a remote counterpart in those cases.

My question is, should **wait_for_connection** support `delegate_to`. As to me one should be testing connections to the target node, not a delegate. IMO, the use-case is sub-optimal (see PR listed).

**Comment by cognifloyd** [comment]
Adding an additional play does not work for me. I use `hosts: all` with a limit on the playbook so that I can target any single host. The purpose of this playbook is to "rename" a host (reconfigure/restart network bits so it shows up as under a different `inventory_host` name). I want to use `wait_for_connection` in a handler to delay ending the play (and consequently the playbook) until the host reconnects.

Using `limit` is great for the first play, but also limits any follow up plays. So, I need `delegate_to` with `wait_for_connection` to use it in a `limit`ed playbook.

**Comment by dagwieers** [comment]
@cognifloyd I don't see what `limit` has to do with anything. If you start a new play with the new targets (so the new `inventory_host` names) it should work out fine.

```yaml
- hosts: original_hosts
  tasks:
  - rename_host:
      old_name: '{{ inventory_hostname }}'
      new_name: '{{ new_name }}'

- hosts: new_hosts
  tasks:
  - wait_for_connection:
```

Obviously it is up to you to create that `new_hosts` list.
Most likely you want to create a new list of hosts in the first play.
(You cannot use `group_by` for obvious reasons)

There might be an alternative by using `set_fact` and changing the `remote_host` hostvar for each target. So that your old target still works, but using the new information. I haven't tried it myself:

```yaml
- hosts: original_hosts
  tasks:
  - rename_host:
      old_name: '{{ inventory_hostname }}'
      new_name: '{{ new_name }}'

  - set_fact:
      remote_host: '{{ new_name }}'

  - wait_for_connection:
```

Ideally the task that renames the host also updates the hostvars entry :-)

**Comment by cognifloyd** [comment]
These are some handlers in a role that validates that a host is identified and connected correctly.
```yaml
- name: Restart Network
  async: 15
  poll: 0
  register: restart_network
  ...

- name: Wait for Reconnect
  listen: Restart Network
  wait_for_connection:
    delay: 5
    connection_timeout: 60
  #delegate_to: correct_inventory_host
  # wait_for_connection does not support delegate_to, so imitate using vars
  vars:
    ansible_host: "{{ hostvars[correct_inventory_host]['ansible_host'] }}"
    ansible_user: "{{ hostvars[correct_inventory_host]['ansible_user'] }}"
    ansible_port: "{{ hostvars[correct_inventory_host]['ansible_port'] }}"
    ansible_ssh_private_key_file: "{{ hostvars[correct_inventory_host]['ansible_ssh_private_key_file'] }}"
    ansible_ssh_common_args: "{{ hostvars[correct_inventory_host]['ansible_ssh_common_args'] }}"
    ansible_python_interpreter: "{{ hostvars[correct_inventory_host]['ansible_python_interpreter'] }}"

- name: Cleanup network restart task
  listen: Restart Network
  #delegate_to: correct_inventory_host
  async_status:
    jid: "{{ restart_network.ansible_job_id }}"
    mode: cleanup
  vars:
    ansible_host: "{{ hostvars[correct_inventory_host]['ansible_host'] }}"
    ansible_user: "{{ hostvars[correct_inventory_host]['ansible_user'] }}"
    ansible_port: "{{ hostvars[correct_inventory_host]['ansible_port'] }}"
    ansible_ssh_private_key_file: "{{ hostvars[correct_inventory_host]['ansible_ssh_private_key_file'] }}"
    ansible_ssh_common_args: "{{ hostvars[correct_inventory_host]['ansible_ssh_common_args'] }}"
    ansible_python_interpreter: "{{ hostvars[correct_inventory_host]['ansible_python_interpreter'] }}"
```
Adding a second play to deal with what cleanly fits within the role's handler seems awkward. It could be done (say when inventory_host != correct_inventory_host) but then the network-related logic gets separated from the role and it becomes harder to follow and debug.

Side note: @dagwieers Thanks for pushing back on my use of limit. Looking into it, I think I have a much better solution for the playbooks where I currently rely on a wide hosts with a pattern in limit. Instead I can do something like: `hosts: "{{ group_name_in_extra_vars }}:&limit_to_this_group_name:!but_not_this_group_name"` A var in a pattern that provides the limit will vastly simplify some of my playbooks.

**Comment by jbemmel** [comment]
I have a use case for delegated wait_for: When the target is only reachable through a bastion host and not directly from the Ansible control host, the reachability check should be executed remotely

**Comment by dagwieers** [comment]
@jbemmel **wait_for** is a (remote) module, so it should work fine with **delegate_to:**. I am not sure why you bring up a reachability check, as **wait_for** doesn't have such a thing AFAIK.

**Comment by juanvalino** [comment]
+1 for this because if the checked host is not directly accessible (e.g you are using a bastion host) the module is useless.

**Comment by jbemmel** [comment]
I use it to wait for SSH to come up, when bringing up a new VM for example ( wait_for port 22 ).
If delegate_to already works then great, I was just responding to your comment "I just don't know if this is desirable as it is counter-intuitive IMO."

In some cases, the VMs I bring up aren't directly reachable from the Ansible host, so I need to be able to execute the wait_for logic on a delegated host that does have reachability

**Comment by dagwieers** [comment]
@juanvalino @jbemmel The **wait_for_connection** action plugin uses the underlying connection method for testing if the connection works (from the Ansible host), so it should work equally well for e.g. SSH using ProxyCommand (which is typically what you would do for bastion hosts).

If the systems are not directly reachable by the Ansible host, then using **wait_for_connection** makes absolutely no sense, because the Ansible host will not be able to continue with those targets anyway. You probably want to use the **wait_for** module in this case?

Again, this is an action plugin, not a module, so it cannot be delegated (i.e. run on another target). The only thing we might be able to do is to test a connection from the Ansible host to another target instead of the configured target, but I don't think that makes sense (it would be confusing or counter-intuitive).

**Comment by Dmitry1987** [comment]
![image](https://user-images.githubusercontent.com/9369080/56696221-12ef8b80-66a0-11e9-9417-0246499a4b82.png)

is it a mistake in the [documentation of ec2 module](https://docs.ansible.com/ansible/latest/modules/ec2_module.html) to use wait_for_connection with delegate_to? 
I tried to see how to specify the target for the wait_for_connection plugin, and found the wait_for which seems a better way to check newly created ec2 host. Should I open a bug for the docs section?

**Comment by ansibot** [comment]
Files identified in the description:
* [`lib/ansible/modules/wait_for_connection.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/wait_for_connection.py)

If these files are incorrect, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by next-jesusmanuelnavarro** [comment]
> ![image](https://user-images.githubusercontent.com/9369080/56696221-12ef8b80-66a0-11e9-9417-0246499a4b82.png)
> 
> is it a mistake in the [documentation of ec2 module](https://docs.ansible.com/ansible/latest/modules/ec2_module.html) to use wait_for_connection with delegate_to?
> I tried to see how to specify the target for the wait_for_connection plugin, and found the wait_for which seems a better way to check newly created ec2 host. Should I open a bug for the docs section?

I think there's a misunderstanding overthere.

I get that *wait_for_connection* is an action plugin and runs from the ansible controller.  It just makes sense once you think of it.

Then, the **intent** of the *delegate_to* is also clear from a user perspective: *"wait for **this** host to be up and running"* (usually within a loop).

Now, the misunderstanding is that the intent is not to *delegate* but to try connectivity (usually within a loop, as I said) **from** the ansible controller, **to** the targets (with the advantage of using any required customizations already in place for the Ansible *"environment"* like proxy/relay, custom SSH port or other SSH connection params, etc.).  That means it shouldn't be implemented by means of *delegate_to* but by means of an optional parameter, say *target_host* or something like that (defaulting to the current host in the play).

##### My use case
Very similar to that from the code above (and, I suppose, quite a common and sensible one):
* within a **hosts: localhost** playbook...
  1. Create a bunch of virtual machines (I'm currently using the *terraform* module on my tests, but it could be any other to same effect)
  1. Refresh inventory (or produce a suitable list of *ansible_hostnames* by any other means)
  1. Wait for all these new hosts to be up & running

...So, the pseudocode would end up being more or less

```yaml
- name: Creates hosts and waits for them to be up & running
  hosts: localhost

  tasks:
  - name: Runs terraform
    terraform:
      # this creates some infrastructure
[...]
  post_tasks:
  - name: Since I use EC2 dynamic inventory, refresh it to "discover" the new infra
   # You'd better don't put this under a when clause, i.e. "when: creation (but not when: destruction)"
   # because of (generalized) bug^H^H^Hfeature request https://github.com/ansible/ansible/issues/41313
    meta: refresh_inventory

  - name: Wait for the infra to be up and running
    wait_for_connection:
      target: "{{ item }}" <- this is the new param
    # delegate_to: "{{ item }}" <- this is wrong! we don't really want to delegate but to target "{{ item }}"
    with_inventory_hostnames:  <- this could be any other list with the desired inventory_hostnames
      - :!ungrouped
```

Hope this makes sense.

**Comment by next-jesusmanuelnavarro** [comment]
> Just target your newly added nodes in a new play and it will actually wait for all your hosts to come online in parallel.

...which means, please correct me if I'm wrong, that you won't be able to encapsulate this very much sensible pattern, *"create infrastructure in some opinionated way and wait for it to be up & running"*, within a role.  Not at least with the help of *wait_for_connection* despite this module being the perfect case for doing it in an *ansible-friendly* way.

**Comment by jborean93** [comment]
I've just tested this out and I cannot reproduce it anymore. A simple `wait_for_connection` with `delegate_to: my-win-host` and it used the connection options of the delegated host as expected.

If this is still a problem for anyone on a newer Ansible version please open a new issue with the reproducer and we can look further into it.
